### PR TITLE
Feat/pending state token permissions handling

### PR DESCRIPTION
**Pull request by osmaczko**
add pending state for token permissions

iterates: https://github.com/status-im/status-desktop/issues/11852

### CODE REVIEW

File: protocol/communities/community_event.go [hunk 1]
```diff
@@ -246,23 +237,19 @@ func (o *Community) updateCommunityDescriptionByCommunityEvent(communityEvent Co
...
 	case protobuf.CommunityEvent_COMMUNITY_MEMBER_TOKEN_PERMISSION_CHANGE:
-		_, exists := o.config.CommunityDescription.TokenPermissions[communityEvent.TokenPermission.Id]
-		if exists {
-			_, err := o.updateTokenPermission(communityEvent.TokenPermission)
-			if err != nil {
-				return err
-			}
-		} else {
-			_, err := o.addTokenPermission(communityEvent.TokenPermission)
+		if o.IsControlNode() {
```
Comment (osmaczko) [inline protocol/communities/community_event.go:240]:
```suggestion
		if o.IsPrivilegedMember(o.MemberIdentity() {
```
Other admins should see pending changes.

File: protocol/communities/community.go [hunk 2]
```diff
@@ -1320,20 +1321,64 @@ func (o *Community) Categories() map[string]*protobuf.CommunityCategory {
...
-func (o *Community) TokenPermissions() map[string]*protobuf.CommunityTokenPermission {
-	return o.config.CommunityDescription.TokenPermissions
+func (o *Community) tokenPermissions() map[string]*CommunityTokenPermission {
+	result := make(map[string]*CommunityTokenPermission, len(o.config.CommunityDescription.TokenPermissions))
+	for _, tokenPermission := range o.config.CommunityDescription.TokenPermissions {
+		result[tokenPermission.Id] = NewCommunityTokenPermission(tokenPermission)
+	}
+
+	// Non-privileged members should not see pending permissions
+	if o.config.EventsData == nil || !o.IsPrivilegedMember(o.MemberIdentity()) {
```
Comment (osmaczko) [inline protocol/communities/community.go:1326]: `!o.IsPrivilegedMember(o.MemberIdentity())` this check shouldn't be needed

TODO: reject token permission events for ordinary members

File: protocol/communities/manager.go [hunk 3]
```diff
@@ -1393,35 +1396,35 @@ func (m *Manager) HandleCommunityEventsMessage(signer *ecdsa.PublicKey, message
...
 	if err != nil {
 		return nil, err
 	}
 
-	// Control node cerifies community events and publish changes
-	// all other nodes only apply changes to the community
-	if changes.Community.IsControlNode() {
-		changes.Community.increaseClock()
-		err = m.persistence.SaveCommunity(changes.Community)
+	// Control node applies events and publish updated CommunityDescription
```
Comment (jrainville) [inline protocol/communities/manager.go:1462]:
```suggestion
	// Control node applies events and publishes updated CommunityDescription
```

File: protocol/communities/community_token_permission.go [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+import (
+	"reflect"
+
+	"github.com/status-im/status-go/protocol/protobuf"
+)
+
+type TokenPermissionState uint8
+
+const (
+	TokenPermissionApproved TokenPermissionState = iota
```
Comment (jrainville) [inline protocol/communities/community_token_permission.go:12]: Do we need a `TokenPermissionRejected` too? Maybe if an admin tries to create a TokenMaster permission or something? (they'd need to hack the client to do that, but still, we need to reject those)

**Review by jrainville — APPROVED** [review]
Nice job

File: protocol/communities/community_token_permission.go (hunk 4 above)
Comment (osmaczko) [reply protocol/communities/community_token_permission.go:12]: I don't think so.

> Maybe if an admin tries to create a TokenMaster permission

Then it will not go through the validation process and be silently ignored by everyone (control node, other admins and members).

One scenario where a token permission event can be explicitly rejected is when control node has newer `CommunityDescription` than the one received in events message. We have a generic solution for that already: https://github.com/status-im/status-go/pull/3876

File: protocol/communities/community_changes.go [hunk 5]
```diff
@@ -79,7 +79,35 @@ func (c *CommunityChanges) HasMemberLeft(identity string) bool {
...
+			changes.TokenPermissionsRemoved[id] = originPermission
+		}
+	}
+
+	// Check for added token permissions
+	for id, permission := range modifiedTokenPermissions {
+		if _, ok := originTokenPermissions[id]; !ok {
+			changes.TokenPermissionsAdded[id] = permission
+		}
+	}
```
Comment (mprakhov) [inline protocol/communities/community_changes.go:104]: why did you remove it from the evaluateCommunityChangesByDescription if this data from CommunityDescription?

File: protocol/communities/community.go [hunk 6]
```diff
@@ -1315,20 +1316,64 @@ func (o *Community) Categories() map[string]*protobuf.CommunityCategory {
...
 
-func (o *Community) TokenPermissions() map[string]*protobuf.CommunityTokenPermission {
-	return o.config.CommunityDescription.TokenPermissions
+func (o *Community) tokenPermissions() map[string]*CommunityTokenPermission {
+	result := make(map[string]*CommunityTokenPermission, len(o.config.CommunityDescription.TokenPermissions))
+	for _, tokenPermission := range o.config.CommunityDescription.TokenPermissions {
+		result[tokenPermission.Id] = NewCommunityTokenPermission(tokenPermission)
+	}
+
+	// Non-privileged members should not see pending permissions
```
Comment (mprakhov) [inline protocol/communities/community.go:1325]: I think this is wrong and tokenPermissions must return all data it has related to the tokenPermissions and not modify the content to return

If permission in a pending state, it must be ignored by privileged members during CommunityEvent handling, wait until Control node handle this event and publish CommunityDescription with the update

**Review by mprakhov — COMMENTED** [review]
Ok, it seems I was late :)

File: protocol/communities/community.go (hunk 6 above)
Comment (osmaczko) [reply protocol/communities/community.go:1325]: https://github.com/status-im/status-go/pull/3925#discussion_r1302011133

File: protocol/communities/community_changes.go (hunk 5 above)
Comment (osmaczko) [reply protocol/communities/community_changes.go:104]: Because permissions are now obtained not only from the community description but also from events. See: https://github.com/status-im/status-go/blob/c3cac93229f25446c6203d4dea40b5d2eae960e9/protocol/communities/community.go#L1319

File: protocol/communities/community.go (hunk 6 above)
Comment (mprakhov) [reply protocol/communities/community.go:1325]: ok, but you already skip it for non-control nodes
https://github.com/status-im/status-go/pull/3925/files#diff-ad79863baafd661e9f1626e2e7ebc0e5db3489ce93c5d5859237446a801b2af2R239-R245

Comment (osmaczko) [reply protocol/communities/community.go:1325]: Permissions are now obtained also from events, this part is skipping only `CommunityDescription` update (`updateCommunityDescriptionByCommunityEvent`).

Comment (osmaczko) [reply protocol/communities/community.go:1325]: > ok, but you already skip it for non-control nodes https://github.com/status-im/status-go/pull/3925/files#diff-ad79863baafd661e9f1626e2e7ebc0e5db3489ce93c5d5859237446a801b2af2R239-R245

Permissions are now obtained also from events, this part is skipping only `CommunityDescription` update (`updateCommunityDescriptionByCommunityEvent`).

Comment (mprakhov) [reply protocol/communities/community.go:1325]: Permissions were obtained from events before, I wrote unit tests for this scenario. All members, who received these permissions, were adding them to their community description. 
From my understanding, your changes should be to ignore from member-side permissions, which are not confirmed by a Control Node.

When you request Community from the database, it should apply all community events on top of the community description. Previously this fn was the same as it was for patching CommunityDescription during handling community events

So you, as a simple member, will not receive patched CommunityDescription with non-approved CommunityPermission. All other admins/token masters must receive it and see it in a pending state
